Fix validation rule to detect tuples in projections or groupbys

juanjux opened the pull request:
Fixes src-d/gitbase/issues/785

Modifies a rule to check if distinct have a tuple argument (`DISTINCT(1,2)`), recursively. Unrelatedly, it also fix ordering in the README for the FROM_BASE64 function (my fault).

There are also a couple unrelated files with `go fmt` changes which I decided to add to this PR since it won't hurt.

Signed-off-by: Juanjo Alvarez <juanjo@sourced.tech>

kuba-- commented on sql/analyzer/optimization_rules.go:
```diff
@@ -1,7 +1,7 @@
 
 import (
-	errors "gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-errors.v1"
```
Actually changes in this file you can discard.

juanjux replied:
It was the result of gofmt over those files which doesn't hurt, but I'll discard to keep the PR tidier.

erizocosmico replied:
This is a change goimports usually does. Instead of having to manually revert it every time I'd just keep it.

kuba-- commented on sql/analyzer/validation_rules.go:
```diff
@@ -3,7 +3,7 @@ package analyzer
 	"strings"
 
-	errors "gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-errors.v1"
```
changes in this file we can also  discard

kuba-- commented on sql/analyzer/resolve_columns.go:
```diff
@@ -38,6 +38,31 @@ func checkAliases(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
 	return n, err
 }
 
+func checkDistinctNoTuples(ctx *sql.Context, a *Analyzer, node sql.Node) (sql.Node, error) {
```
I don't know if it makes sense here, but did you test with wrapped nodes or subqueries?
Maybe we need to inspect distinct nodes recursively here, e.g. 
https://github.com/src-d/go-mysql-server/blob/master/sql/analyzer/assign_indexes.go#L43

erizocosmico commented:
It should not only be distinct, but anything that is projected. A tuple simply can't be projected (neither in Project nor GroupBy)

juanjux commented:
@erizocosmico change to detect tuples in projections and groupbys implemented, PTAL when you can.

erizocosmico commented on sql/analyzer/resolve_columns.go:
```diff
@@ -38,6 +38,41 @@ func checkAliases(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
+
+	a.Log("check no tuples as in projection, node of type: %T", node)
+	return node.TransformUp(func(node sql.Node) (sql.Node, error) {
+		project, ok := node.(*plan.Project)
```
Project and GroupBy share a common interface: sql.Expressioner, which returns all expressions in a node.
You could leverage that common interface to reduce the code of this function to the following:

```go
switch node := node.(type) {
case *plan.Project, *plan.GroupBy:
        for _, e := range node.(sql.Expressioner).Expressions() {
                if sql.IsTuple(e.Type()) {
                        return nil, ErrTupleProjected.New()
                }
        }
}
```

Note that in this snippet I added a fix to your code. Not always you will have an `expression.Tuple` at the top level of an expression. The safe way to check if an expression is a tuple is if its type is a tuple. For this, we have a helper function named `sql.IsTuple`, so you can do `sql.IsTuple(expr.Type())`.

juanjux replied:
Done (in the existing rule that wasn't doing it that way).

erizocosmico commented on sql/analyzer/resolve_columns_test.go:
```diff
@@ -75,6 +75,64 @@ func TestMisusedAlias(t *testing.T) {
+	})
+
+	node := plan.NewProject([]sql.Expression{
+		expression.NewTuple(
```
We should add a new test case in which this `expression.NewTuple` is wrapped in a `expression.NewAlias`, for example. With that case, the current implementation would fail, but I already suggested a fix to that in my previous comment.

juanjux replied:
Done (in the testcase for the existing rule).

erizocosmico commented on sql/analyzer/rules.go:
```diff
@@ -66,4 +67,5 @@ var (
 	// ErrMisusedAlias is returned when a alias is defined and used in the same projection.
 	ErrMisusedAlias = errors.NewKind("column %q does not exist in scope, but there is an alias defined in" +
 		" this projection with that name. Aliases cannot be used in the same projection they're defined in")
+	ErrTupleProjected = errors.NewKind("unexpected tuple found, maybe remove the ()?")
```
I would change the message to `unexpected tuple found in SELECT clause, tuples cannot be selected`. Because not always a tuple will be a tuple only as a result of having ().

juanjux replied:
:+1:

juanjux commented:
Update: there was already a validation rule checking that there are not tuples in Projections or GroupBys (https://github.com/src-d/go-mysql-server/blob/master/sql/analyzer/validation_rules.go#L198), it even have tests with tuples (https://github.com/src-d/go-mysql-server/blob/master/sql/analyzer/validation_rules_test.go#L234) but adding a test with a DISTINCT->Projection->Tuple doesn't detect the error (and it crash like the issue reported if you do it on the REPL), so I'm not investigating why this existing rule doesn't work for this case.

juanjux commented:
Yes, with a recursive inspection the existing rule works beautifully, PTAL again since I removed basically everything from the previous commits and rewrote again.
